### ISSUE TITLE
Augmentation strength always increasing - what is the rt heuristic doing?

### ISSUE THREAD

**Issue by 100330706** [NVlabs/stylegan2-ada#27]
Can someone give me some insight about the heuristic value (rt) that controls the augmentation strength in the ada pipeline? In augments.py the code is the following:

`strength += nimg_ratio * np.sign(rt - self.tune_target)`

I am training using the default tune_target (0.6) and I've realized that strength is always increasing (as rt > tune_target most of the times). I'm close to an augmentation strength of 10 after 1000 kimg or so. What does exactly mean this? What is the reasoning behind rt? Does the augmentation strength increasingly grow because my dataset is not good enoguh so the network feels like the strength needs to keep increasing? Thanks in advance.

**Comment by mayankgrwl97** [comment]
Facing the same issue, would be great if someone could explain. Shouldn't the augmentation strength be <= 1?

**Comment by 100330706** [comment]
I'm thinking about just clipping the value so the maximum is 1. However this won't solve the fact that the heuristics are detecting insane overfitting because otherwise it woudn't be increasing :(

**Comment by woctezuma** [comment]
I believe that this indicate an imbalance between the discriminator and the generator during the training.

What I would do is increase the value of `tune_kimg`:

https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/training/augment.py#L34

This will decrease the value of `nimg_ratio`:

https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/training/augment.py#L110

which in turns will decrease the value of the delta of strength:

```python
strength += nimg_ratio * np.sign(rt - self.tune_target)
```
or
```python
strength += nimg_ratio * np.sign(rv - self.tune_target)
```
in:
https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/training/augment.py#L113-L123

**Comment by nikhilweee** [comment]
From what I understand, and from what's listed in Section 3 of the [paper](https://arxiv.org/abs/2006.06676), `rt` is simply the expected value of the scores from the discriminator. It's used as a measure of overfitting where `rt = 0` means no overfitting and `rt = 1` indicates complete overfitting. With the default parameters, the code tries to achieve a target value of `rt = 0.6` for the `ada` configuration. 
>The second heuristic, rt, estimates the portion of the training set that gets positive discriminator outputs. For both heuristics, r = 0 means no overfitting and r = 1 indicates complete overfitting, and our goal is to adjust the augmentation probability p so that the chosen heuristic matches a suitable target value.

The way it does so is by continuously adjusting the augmentation strength `strength += nimg_ratio * np.sign(rt - self.tune_target)`. Essentially, whenever `rt` is above the target value, the augmentation strength is increased and whenever `rt` is below the target value, the augmentation strength is decreased in an attempt to bring `rt` closer to the target value. 
>We control the augmentation strength p as follows. We initialize p to zero and adjust its value once every four minibatches based on the chosen overfitting heuristic. If the heuristic indicates too much/little overfitting, we counter by incrementing/decrementing p by a fixed amount. We set the adjustment size so that p can rise from 0 to 1 sufficiently quickly, e.g., in 500k images.  

https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/training/augment.py#L113-L116

But as you can imagine, there has to be a limit to the augmentation strength because too many augmentations will lead to leaky results, and too few augmentations (where `strength < 0`) do not make sense. As a matter of fact, there _is_ a lower bound on the minimum value for the augmentation strength, and it is capped to zero.  
>  After every step we clamp p from below to 0. We call this variant adaptive discriminator augmentation (ADA).

https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/training/augment.py#L133 

However, there is no upper bound on the augmentation strength, which is why it keeps on increasing.

**Comment by nikhilweee** [comment]
Personally, I see three different ways to counter this problem, listed below in order of preference:

1. Slow down updates to the augmentation strength (as suggested by @woctezuma in https://github.com/NVlabs/stylegan2-ada/issues/27#issuecomment-723445213)
Change this value from `100` to `500` or `1000` or whatever you like:
https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/train.py#L399

2. Force an upper bound on the augmentation strength
Add something like `strength = min(strength, 0.9)` to the following: https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/training/augment.py#L133 

3. Adjust the desired target value of `rt`
Use `--target 0.5` or `--target 0.55` while calling the training script `train.py`

I would choose the first or the second approach over the third one because the [paper](https://arxiv.org/abs/2006.06676) suggests that `rt` is not affected as much by hyperparameters. So changing `--target` would not make that much of a difference.
>The second heuristic, rt, estimates the portion of the training set that gets positive discriminator outputs. We have found this to be far less sensitive to the chosen target value and other hyperparameters than the obvious alternative of looking at E[Dtrain] directly.

**Comment by nikhilweee** [comment]
> Facing the same issue, would be great if someone could explain. Shouldn't the augmentation strength be <= 1?

@mayankgrwl97 I believe that the augmentation probability is practically capped to a maximum value of 1 because of `L171`. 
https://github.com/NVlabs/stylegan2-ada/blob/8c3fd8bac3e5a54a20e860fc163d6e67cc03c623/training/augment.py#L169-L173
So an ever increasing value of the augmentation strength on the tensorboard (tag `Progress/augment`) probably indicates that despite increasing the augmentation probability to the fullest, `rt` is still too far from the target value, which in turn suggests that the training has collapsed.

**Comment by sangyun884** [comment]
![image](https://user-images.githubusercontent.com/71681194/106090478-9986a600-616d-11eb-8b6a-3f0b7d46ca96.png)

Increment of p is normal according to original paper. I think set the upper bound of p(0.8, for example) is needed to prevent leaking.
